### PR TITLE
Keysyms: explicit bounds

### PR DESCRIPTION
**Pull request by wismill**
Working on a fix of #413, I realize that some bounds related should be made explicit. 

I added `XKB_KEYSYM_NAME_MAX_SIZE` to the public API and a keysym iterator API for internal use.

### CODE REVIEW

File: src/keysym.c [hunk 1]
```diff
@@ -55,6 +55,31 @@
...
+        int32_t mid = (lo + hi) / 2;
+        if (ks > keysym_to_name[mid].keysym) {
+            lo = mid + 1;
+        } else if (ks < keysym_to_name[mid].keysym) {
+            hi = mid - 1;
+        } else {
+            return true;
+        }
+    }
+    return false;
```
Comment (whot) [inline src/keysym.c:96]: I keep looking at this and thinking there's a way to re-use this binary search in `xkb_keysym_get_name()`. Maybe through a combination of:
```c
static ssize_t find_offset(xkb_keysym_t ks); /* returns -1 on error or the offset */

bool xkb_keysym_is_assigned(xkb_keysym_t ks) {
    return find_offset(ks) != -1; 
}

int
xkb_keysym_get_name() {
     ssize_t offset = find_offset(ks);
     if (offset != -1)
         return snprintf(get_name(&keysym_to_name[offset]));
     ...
}
```

File: src/keysym.h [hunk 2]
```diff
@@ -72,9 +74,31 @@
 /** Maximum Unicode keysym, correspoding to the maximum Unicode code point */
 #define XKB_KEYSYM_UNICODE_MAX    0x0110ffff
 
+#ifdef ENABLE_PRIVATE_APIS
```
Comment (whot) [inline src/keysym.h:77]: am I missing something here? `src/keysym.h` is an internal header anyway so we don't need a special `#ifdef` to keep the APIs private, right?

I tried removing this `#ifdef` here and `nm` doesn't list any of the symbols so we're good here.

File: src/keysym.c [hunk 3]
```diff
@@ -61,6 +61,130 @@ get_name(const struct name_keysym *entry)
...
+    uint32_t index;
+    xkb_keysym_t keysym; /* Current keysym */
+    bool explicit;       /* If true, explicitly named keysym */
+};
+
+struct xkb_keysym_iterator*
+xkb_keysym_iterator_new(bool unicode)
+{
+    struct xkb_keysym_iterator*
+    iter = calloc(1, sizeof(*iter));
```
Comment (whot) [inline src/keysym.c:110]:
```suggestion
    struct xkb_keysym_iterator* iter = calloc(1, sizeof(*iter));
```

**Review by whot — COMMENTED** [review]
Mostly LGTM, the one complaint I have is the `XKB_KEYSYM_NAME_MAX_SIZE` define that will be part of the ABI and likely lead to application failures (much further) down the road.

Otherwise, the usual minor suggestions but as usual, this is great work, thanks!

File: src/keysym.c [hunk 4]
```diff
@@ -61,6 +61,130 @@ get_name(const struct name_keysym *entry)
...
+    iter->index = 0;
+    iter->keysym = keysym_to_name[0].keysym;
+    return iter;
+}
+
+void
+xkb_keysym_iterator_unref(struct xkb_keysym_iterator *iter)
+{
+    free(iter);
+}
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.c:122]: Done

File: src/keysym.c [hunk 5]
```diff
@@ -61,6 +61,130 @@ get_name(const struct name_keysym *entry)
...
+
+xkb_keysym_t
+xkb_keysym_iterator_keysym(struct xkb_keysym_iterator *iter) {
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.c:125]: Done

File: src/keysym.c [hunk 6]
```diff
@@ -61,6 +61,130 @@ get_name(const struct name_keysym *entry)
...
+    return iter->keysym;
+}
+
+int
+xkb_keysym_iterator_name(struct xkb_keysym_iterator *iter,
+                         char *buffer, size_t size) {
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.c:131]: Done

File: src/keysym.c [hunk 7]
```diff
@@ -61,6 +61,130 @@ get_name(const struct name_keysym *entry)
...
+
+/* Iterate over the *assigned* keysyms.
+ *
+ * Use:
+ *
+ * ```c
+ * struct xkb_keysym_iterator *iter = xkb_keysym_iterator_new(true);
+ * do {
+ *   ...
+ * } while (xkb_keysym_iterator_next(iter));
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.c:150]: Done

File: src/keysym.c (hunk 3 above)
Comment (wismill) [reply src/keysym.c:110]: Done

File: src/keysym.c [hunk 8]
```diff
@@ -61,6 +61,130 @@ get_name(const struct name_keysym *entry)
...
+    bool explicit;       /* If true, explicitly named keysym */
+};
+
+struct xkb_keysym_iterator*
+xkb_keysym_iterator_new(bool unicode)
+{
+    struct xkb_keysym_iterator*
+    iter = calloc(1, sizeof(*iter));
+    iter->unicode = unicode;
+    iter->explicit = !unicode;
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.c:112]: Yes, it is confusing!

Refactored:

- The keysym property `explicit` is now computed on-demand by `xkb_keysym_iterator_is_explicitly_named`. It returns true iff the keysym is explicitly named, independently if in Unicode range or not. Actually it is not used… do you have a test in mind? Else I should probably just remove it.
- The iterator config `unicode` is replaced by `explicit_only` with the opposite behaviour: `xkb_keysym_iterator_new(explicit=true)` iterates _only_ over explicitly named keysyms, and with `explicit=false` it iterates over all assigned keysyms (explicit + Unicode range).

I hope it is clearer.

**Review by wismill — COMMENTED** [review]
@whot I refactored the code and fixed the commits. There are still some comments to address.

File: src/keysym.h [hunk 9]
```diff
@@ -57,13 +57,24 @@
  */
 /** Minimum keysym value */
 #define XKB_KEYSYM_MIN            0x00000000
+/** Minimum keysym value assigned */
+ #define XKB_KEYSYM_MIN_ASSIGNED  0x00000000
```
[the remark this answers is not in the document]
Comment (wismill) [reply src/keysym.h:64]: I have no objection, but what is the added value?

File: include/xkbcommon/xkbcommon.h.jinja [hunk 10]
```diff
@@ -0,0 +1,2002 @@
+/*
```
[the remark this answers is not in the document]
Comment (wismill) [reply include/xkbcommon/xkbcommon.h.jinja:1]: Fixed

File: include/xkbcommon/xkbcommon.h.jinja [hunk 11]
```diff
@@ -442,15 +451,16 @@ struct xkb_rule_names {
  * @param[in]  size   Size of the buffer.
  *
  * @warning If the buffer passed is too small, the string is truncated
- * (though still NUL-terminated); a size of at least 64 bytes is recommended.
+ * (though still NUL-terminated); a size of at least
+ * `XKB_KEYSYM_NAME_MAX_SIZE` bytes is recommended.
```
[the remark this answers is not in the document]
Comment (wismill) [reply include/xkbcommon/xkbcommon.h.jinja:455]: `XKB_KEYSYM_MAX_SIZE` is computed from the header file, so it will always be the minimal correct value to use. The only issue I see that could happen is it if an app is compiled with an xkbcommon version different from the one it is dynamically linked, _and_ this version has a greater value for `XKB_KEYSYM_NAME_MAX_SIZE`.

But I still think it is better to avoid unnamed numeric constants.

Possible solutions:
- Add a margin: e.g. `XKB_KEYSYM_NAME_MAX_SIZE = <current exact value> + 15`
- Set to current advised length `XKB_KEYSYM_NAME_MAX_SIZE = 64`.

[the remark this answers is not in the document]
Comment (whot) [reply include/xkbcommon/xkbcommon.h.jinja:455]: > The only issue I see that could happen is it if an app is compiled with an xkbcommon version different from the one it is dynamically linked, and this version has a greater value 

yes, that's the one I'm concerned about because that's pretty much the case for any application that uses a shared library but doesn't get updated regularly.

> But I still think it is better to avoid unnamed numeric constants.

So the thing here is that it's not a numeric constant, it's a buffer size that the application gets to pick.

File: src/keysym.h (hunk 9 above)
[the remark this answers is not in the document]
Comment (whot) [reply src/keysym.h:64]: readability, primarily ("oh, this zero is a keysym!") but also *some* typesafety because `xkb_keysym_t` is a `uint32_t` whereas `0x0` is a plain boring `int`.

File: scripts/update-headers.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+    args.root,
+    Path("include/xkbcommon/xkbcommon.h"),
+)
+
+generate(
+    jinja_env,
+    keysyms_data,
+    args.root,
+    Path("src/keysym.h"),
+)
```
[the remark this answers is not in the document]
Comment (whot) [reply scripts/update-headers.py:104]: Can we do this instead then:
```python
for file in ("src/keysym.h", "include/xkbcommon/xkbcommon.h"):
    generate(..., Path(file))
```

This makes it immediately visible that the two files are processed the same way.

File: include/xkbcommon/xkbcommon.h.jinja (hunk 11 above)
[the remark this answers is not in the document]
Comment (whot) [reply include/xkbcommon/xkbcommon.h.jinja:455]: oh, and ftr: if you want this as a *private* `#define` in some utility header for use by our tests/tools that's perfectly fine, I just really dislike this as part of the header API.

Oh, and for explanation: this is a bug that won't hit us for years, if ever. But when it hits we then have several years of applications that are all broken in the same way and hard to fix.

File: src/utils.h [hunk 13]
```diff
@@ -43,9 +43,11 @@
 #ifdef _WIN32
 # include <direct.h>
 # include <io.h>
+# include <BaseTsd.h>
 # ifndef S_ISDIR
 #  define S_ISDIR(m) (((m) & S_IFMT) == S_IFDIR)
 # endif
+typedef SSIZE_T ssize_t;
```
Comment (whot) [inline src/utils.h:50]: does this one need to be squashed in somewhere or is it a standalone one? If the latter could you factor it out into a separate PR and merge it directly please? If the former, please squash it in :)

File: src/keysym.c [hunk 14]
```diff
@@ -122,6 +128,98 @@ xkb_keysym_is_assigned(xkb_keysym_t ks)
...
 }
 
+struct xkb_keysym_iterator {
+    bool explicit;       /* If true, traverse only explicitly named keysyms */
+    int32_t index;       /* Current index in `keysym_to_name` */
+    xkb_keysym_t keysym; /* Current keysym */
+};
+
+struct xkb_keysym_iterator*
+xkb_keysym_iterator_new(bool explicit)
```
Comment (whot) [inline src/keysym.c:138]: How about `iterate_only_explicit_syms`? It's only ever typed here once anyway so having an awkwardly long name means it self-documents :)

File: src/keysym.c [hunk 15]
```diff
@@ -122,6 +128,98 @@ xkb_keysym_is_assigned(xkb_keysym_t ks)
...
+}
+
+xkb_keysym_t
+xkb_keysym_iterator_get_keysym(struct xkb_keysym_iterator *iter)
+{
+    return iter->keysym;
+}
+
+inline bool
+xkb_keysym_iterator_is_explicitly_named(struct xkb_keysym_iterator *iter)
```
Comment (whot) [inline src/keysym.c:161]: skip the `inline` here please, it won't be inlined if it's a public function anyway. And `inline` is usually ignored anyway by the compiler, I mostly use it as `static inline` to stop the compiler from complaining about unused functions in headers.

File: src/keysym.c (hunk 8 above)
[the remark this answers is not in the document]
Comment (whot) [reply src/keysym.c]: yep, thanks for the refactor (and the explanation).

`xkb_keysym_iterator_is_explicitly_named` *looks* like it'll be useful and it's a cheap one, so IMO we can leave it in place.

File: scripts/update-headers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (wismill) [reply scripts/update-headers.py:104]: Agreed, but needed anymore now that I removed `XKB_KEYSYM_NAME_MAX_SIZE` from public API.

File: src/keysym.h (hunk 2 above)
Comment (wismill) [reply src/keysym.h:80]: 416 bytes on my machine. Not much, but I guess it is still good to keep the size low, especially if it does not require effort. I would like to investigate the cache misses some day.

File: include/xkbcommon/xkbcommon.h.jinja (hunk 11 above)
[the remark this answers is not in the document]
Comment (wismill) [reply include/xkbcommon/xkbcommon.h.jinja:455]: I see, thank you to take the time to explain. So I refactored this to be used internally only.

**Review by wismill — COMMENTED** [review]
Fix comment review and rebased.
